src/main/java/com/atlassian/bitbucket/jenkins/internal/scm/BitbucketSCMSource.java

```diff
@@ -332,77 +277,71 @@ protected void retrieve(@CheckForNull SCMSourceCriteria criteria, SCMHeadObserve
-                        serverId, repository.getMirroringDetails().getMirrorName());
-        initialize(cloneUrl, bitbucketSCMRepository);
+        BitbucketServerConfiguration serverConfiguration = mayBeServerConf.get();
+        String cloneUrl, selfLink;
+
+        BitbucketScmHelper scmHelper = descriptor.getBitbucketScmHelper(serverConfiguration.getBaseUrl(),
+                getCredentials().orElse(null));
+
+        if (repository.isMirrorConfigured()) {
+            EnrichedBitbucketMirroredRepository fetchedRepository = descriptor.createMirrorHandler(scmHelper)
+                    .fetchRepository(
+                            new MirrorFetchRequest(
+                                    serverConfiguration.getBaseUrl(),
+                                    getOwner(),
+                                    getCredentialsId(),
+                                    getProjectName(),
+                                    getRepositoryName(),
+                                    getMirrorName()));
+            BitbucketRepository underlyingRepo = fetchedRepository.getRepository();
+            repository = new BitbucketSCMRepository(getCredentialsId(), getSshCredentialsId(),
+                    underlyingRepo.getProject().getName(), underlyingRepo.getProject().getKey(), underlyingRepo.getName(),
+                    underlyingRepo.getSlug(), getServerId(), fetchedRepository.getMirroringDetails().getMirrorName());
+            cloneUrl = underlyingRepo.getCloneUrl(repository.getCloneProtocol()).map(BitbucketNamedLink::getHref).orElse("");
+            selfLink = fetchedRepository.getRepository().getSelfLink();
+        } else {
+            BitbucketRepository fetchedRepository = scmHelper.getRepository(getProjectName(), getRepositoryName());
+            repository = new BitbucketSCMRepository(getCredentialsId(), getSshCredentialsId(),
+                    fetchedRepository.getProject().getName(), fetchedRepository.getProject().getKey(),
+                    fetchedRepository.getName(), fetchedRepository.getSlug(), getServerId(), "");
+            cloneUrl = fetchedRepository.getCloneUrl(repository.getCloneProtocol()).map(BitbucketNamedLink::getHref).orElse("");
+            selfLink = fetchedRepository.getSelfLink();
+        }
+        // Self link contains `/browse` which we must trim off.
+        selfLink = selfLink.substring(0, max(selfLink.lastIndexOf("/browse"), 0));
+        if (isBlank(cloneUrl)) {
+            LOGGER.info("No clone url found for repository: " + repository.getRepositoryName());
+        }
+
+        // Initialize the Git SCM source
+        UserRemoteConfig remoteConfig = new UserRemoteConfig(cloneUrl, repository.getRepositorySlug(), null,
+                repository.getCloneCredentialsId());
+        gitSCMSource = new CustomGitSCMSource(remoteConfig.getUrl(), repository);
+        gitSCMSource.setBrowser(new Stash(selfLink));
```

Comment:
Should this now be a `BitbucketServer` instead? I believe we've upgraded to a supporting git plugin version? (no need to change here, but we should probably do that change)

Reply:
This seems as sensible a PR to do it in as any :)